llvm/lib/TargetParser/Host.cpp

```diff
@@ -2002,6 +2003,76 @@ bool sys::getHostCPUFeatures(StringMap<bool> &Features) {
 
   return true;
 }
+#elif defined(__linux__) && defined(__riscv)
+// struct riscv_hwprobe
+struct RISCVHwProbe {
+  int64_t Key;
+  uint64_t Value;
+};
+bool sys::getHostCPUFeatures(StringMap<bool> &Features) {
+  RISCVHwProbe Query[]{{/*RISCV_HWPROBE_KEY_BASE_BEHAVIOR=*/3, 0},
+                       {/*RISCV_HWPROBE_KEY_IMA_EXT_0=*/4, 0}};
+  int Ret = syscall(/*__NR_riscv_hwprobe=*/258, /*pairs=*/Query,
+                    /*pair_count=*/std::size(Query), /*cpu_count=*/0,
+                    /*cpus=*/0, /*flags=*/0);
+  if (Ret != 0)
+    return false;
+
+  uint64_t BaseMask = Query[0].Value;
+  // Check whether RISCV_HWPROBE_BASE_BEHAVIOR_IMA is set.
+  if (BaseMask & 1) {
+    Features["i"] = true;
+    Features["m"] = true;
+    Features["a"] = true;
+  }
+
+  uint64_t ExtMask = Query[1].Value;
```

Comment:
If the base behavior isn't `RISCV_HWPROBE_BASE_BEHAVIOR_IMA`, can we still use the value in `RISCV_HWPROBE_KEY_IMA_EXT_0`? It has IMA in its name.

Reply:
I think this is not safe. The minimal ISA for RISC-V Linux is IMA. If someone modified the kernel to run on some CPUs without M or A extension(theoretically A can be replaced with Zalrsc), the base behavior probe will be zero but it doesn’t mean we have IMA.

Reply:
FYI, openjdk still uses the value in `RISCV_HWPROBE_KEY_IMA_EXT_0` even if `RISCV_HWPROBE_BASE_BEHAVIOR_IMA` is not set.

See https://github.com/openjdk/jdk/blob/e527e1c32fcc7b2560cec540bcde930075ac284a/src/hotspot/os_cpu/linux_riscv/riscv_hwprobe.cpp#L142-L185.

Reply:
I think this is likely to be fine in practice.  

From https://docs.kernel.org/arch/riscv/hwprobe.html, I see "If a key is unknown to the kernel, its key field will be cleared to -1, and its value set to 0. "  If I'm reading that properly, if RISCV_HWPROBE_KEY_IMA_EXT_0 isn't a valid key, it is the kernel's responsible for clearing all bits to zero, so this code should work.